Initial gin-gonic instrumentation with otel-go

krnowak opened the pull request:
This is some initial code that does the instrumentation. Still need to fix the tests, add docs and a Dockerfile with an example app. So it's work in progress.

There is the `common.go` file that is mostly about attributes for the span. I suppose it will end up being elsewhere in the long run.

jmacd commented on plugins/gin-gonic/gin/common.go:
```diff
@@ -0,0 +1,226 @@
+// Copyright The OpenTelemetry Authors
```
I take it you envision this being a common module of code shared by plugins. It might also be used in the built-in plugins included with the base library, right? There's a precedent in `sdk/resource/resourcekeys/const.go`, but I think these should be defined in the API not the SDK package tree.

krnowak replied:
Yeah, but I'd like to have this code to be used by more before trying to put it into otel-go. For now it can be a shared module within otel-go-contrib.

MrAlias commented on plugins/gin-gonic/gin/example_test.go:
```diff
@@ -0,0 +1,69 @@
+// Copyright 2016-2020 Datadog, Inc.
+// Copyright The OpenTelemetry Authors
```
Traditionally I've seen this the other way around:

```
// Copyright The OpenTelemetry Authors
// Copyright 2016-2020 Datadog, Inc.
``` 

INAL, but I think this comes from the idea that the original license needs to remain unmodified if copied.

Also, [the NOTICES file](https://github.com/DataDog/dd-trace-go/blob/8fb554ff7cf694267f9077ae35e27ce4689ed8b6/NOTICE) from the originating repo needs to be included.

Is there a specific reason we are including this code instead of writing our own?

andrewhsu replied:
Because the original source that @krnowak used as reference has dual license choice: either mit or al2, the [mit](https://opensource.org/licenses/MIT) was chosen because it is a lot simpler and does not require the NOTICES file: https://github.com/DataDog/dd-trace-go/pull/571#issue-364984406

As for the order, I think it is ok to have opentelemetry as the second item. This PR conversation can serve as the record for how the code was derived.

MrAlias replied:
@andrewhsu: The code this was copied from was not licensed under an MIT license. It is dual licensed as Apache2 or BSD3. If we want to use the BSD3 that license will need to be copied into this file (part of the explicit requirement to reuse code with that license type), otherwise we'll need to preserve the original Apache2 copyright (including the notices file and the header):

> 4. Redistribution. You may reproduce and distribute copies of the Work or Derivative Works thereof in any medium, with or without modifications, and in Source or Object form, provided that You meet the following conditions:
> 
>    1. You must give any other recipients of the Work or Derivative Works a copy of this License; and
>    1. You must cause any modified files to carry prominent notices stating that You changed the files; and
>    1. You must retain, in the Source form of any Derivative Works that You distribute, all copyright, patent, trademark, and attribution notices from the Source form of the Work, excluding those notices that do not pertain to any part of the Derivative Works; and
>    1. If the Work includes a "NOTICE" text file as part of its distribution, then any Derivative Works that You distribute must include a readable copy of the attribution notices contained within such NOTICE file, excluding those notices that do not pertain to any part of the Derivative Works, in at least one of the following places: within a NOTICE text file distributed as part of the Derivative Works; within the Source form or documentation, if provided along with the Derivative Works; or, within a display generated by the Derivative Works, if and wherever such third-party notices normally appear. The contents of the NOTICE file are for informational purposes only and do not modify the License. You may add Your own attribution notices within Derivative Works that You distribute, alongside or as an addendum to the NOTICE text from the Work, provided that such additional attribution notices cannot be construed as modifying the License.
> 
> You may add Your own copyright statement to Your modifications and may provide additional or different license terms and conditions for use, reproduction, or distribution of Your modifications, or for any such Derivative Works as a whole, provided Your use, reproduction, and distribution of the Work otherwise complies with the conditions stated in this License.

andrewhsu replied:
@MrAlias apologies, you're right. even the PR i linked to in my original comment says BSD3. 🤦 i think i used the wrong link from the wrong tab when composing the comment too quickly. also, kids: don't do drugs.

In that case, @krnowak I recommend we adhere to the AL2 since this repo is already setup up for it and with these changes to this PR:

1. add to the top of the source file a code comment that says `// Code initially derived from https://github.com/DataDog/dd-trace-go/blob/0b8240f0a30d8339b3250ff20d888f40d99759a7/contrib/gin-gonic/gin/example_test.go` or whatever is the appropriate source link
2. add `NOTICES` file to top-level of the repo with contents from https://github.com/DataDog/dd-trace-go/blob/0b8240f0a30d8339b3250ff20d888f40d99759a7/NOTICE

andrewhsu replied:
For reference, I found an example of derived source that cites original source in the kubernetes project: https://github.com/kubernetes/kubernetes/blob/b19de7f9b6a23f669335a10cb1c0e8de10324a78/pkg/credentialprovider/azure/azure_acr_helper.go#L43

krnowak replied:
@andrewhsu: The first point is already fulfilled, I write something like "// Based on https://github.com/DataDog/dd-trace-go/blob/8fb554ff7cf694267f9077ae35e27ce4689ed8b6/contrib/gin-gonic/gin/gintrace_test.go"

Will add the NOTICES file.

krnowak commented:
Ready for review. For now I have added a mock tracer to otel-go-contrib instead of exposing the one in otel-go. This is to avoid waiting for the potential otel-go changes to make it to master.

lizthegrey commented:
Didn't Datadog sign the CLA allowing the copyright notice to read just the OTel authors?

krnowak commented:
Updated:

- `plugin/gin-gonic/gin/common{,_test}.go` are now `internal/trace/http{,_test}.go`
- fixed an import path comment in `plugin/gin-gonic/gin/doc.go` to use the vanity path

Also me when you talk about legal stuff: <insert meme with Homer Simpson backing into the bushes> ;)

krnowak commented:
Updated again to fix a documentation of the `WithTracer` option.

andrewhsu commented:
> Didn't Datadog sign the CLA allowing the copyright notice to read just the OTel authors?

@lizthegrey i don't believe anybody from datadog had committed to this repo with a signed CLA as was done in the opentelemetry-python-contrib repo https://github.com/open-telemetry/opentelemetry-python-contrib/pull/1

krnowak commented:
Updated, did some small fixes in tests, and added the NOTICE file.

andrewhsu commented:
> I'm sorry, but I'm going to need to block merging this until we can discuss the CLA, the copyright notice, etc. at governance committee.

@lizthegrey is there an eta on discussing with the governance committee?

fyi i used https://github.com/cncf/foundation/blob/master/copyright-notices.md#what-about-third-party-code as guidance for this PR since the code is altered from original source

lizthegrey commented:
> > I'm sorry, but I'm going to need to block merging this until we can discuss the CLA, the copyright notice, etc. at governance committee.
> 
> @lizthegrey is there an eta on discussing with the governance committee?

Tomorrow.

> fyi i used [cncf/foundation:copyright-notices.md@`master`#what-about-third-party-code](https://github.com/cncf/foundation/blob/master/copyright-notices.md#what-about-third-party-code) [](/cncf/foundation/blob/HEAD@{2020-04-22T21:04:43Z}/copyright-notices.md#what-about-third-party-code) as guidance for this PR since the code is altered from original source

There are certain sensitivities here about not wanting to bias towards any one vendor in the copyright notices, etc. -- none of the other code contributed to OTel says "copyright Honeycomb", "copyright Lightstep", etc. so it's a matter of fairness.

lizthegrey commented:
> > > I'm sorry, but I'm going to need to block merging this until we can discuss the CLA, the copyright notice, etc. at governance committee.
> > 
> > 
> > @lizthegrey is there an eta on discussing with the governance committee?
> 
> Tomorrow.
> 
> > fyi i used [cncf/foundation:copyright-notices.md@`master`#what-about-third-party-code](https://github.com/cncf/foundation/blob/master/copyright-notices.md#what-about-third-party-code) [](/cncf/foundation/blob/HEAD@{2020-04-22T21:10:52Z}/copyright-notices.md#what-about-third-party-code)  as guidance for this PR since the code is altered from original source
> 
> There are certain sensitivities here about not wanting to bias towards any one vendor in the copyright notices, etc. -- none of the other code contributed to OTel says "copyright Honeycomb", "copyright Lightstep", etc. so it's a matter of fairness.

We met. @tedsuo will follow up with more specific detail, but the CNCF third party code applies _unless_ the contributor has waived the requirement. In this case, the contributor Datadog _has_ waived the requirement. One way to move forward may be to explicitly get a Datadog engineer to append the commit that strips the copyright notices to make clear that it's officially sanctioned.

tedsuo commented:
Thanks @lizthegrey. I'm reaching out to Datadog. I feel we have been given permission to change these licenses, but I will confirm again, and look to adding a clause in our community repo to clarify this subject.

jeremy-lq commented:
Datadog did give our explicit permission to attribute copyright to "OpenTelemetry Authors" here: https://github.com/open-telemetry/community/issues/305#issuecomment-603537422

I wasn't previously aware of this thread. If there are future questions about the topic or clarification needed, do feel free to mention me so I can take a look.

lizthegrey commented:
Thanks, y'all.
